listenbrainz/webserver/views/stats_api.py

```diff
@@ -498,7 +508,127 @@ def get_daily_activity(user_name: str):
+    :type range: ``bool``
+    :statuscode 200: Successful query, you have data!
+    :statuscode 204: Statistics for the user haven't been calculated, empty response will be returned
+    :statuscode 400: Bad request, check ``response['error']`` for more details
+    :statuscode 404: User not found
+    :resheader Content-Type: *application/json*
+
+    """
+    user = db_user.get_by_mb_id(user_name)
+    if user is None:
+        raise APINotFound("Cannot find user: {}".format(user_name))
+
+    stats_range = request.args.get('range', default='all_time')
+    if not _is_valid_range(stats_range):
+        raise APIBadRequest("Invalid range: {}".format(stats_range))
+
+    recalculate_param = request.args.get('force_recalculate', default='false')
+    if recalculate_param.lower() not in ['true', 'false']:
+        raise APIBadRequest("Invalid value of force_recalculate: {}".format(recalculate_param))
+    force_recalculate = recalculate_param.lower() == 'true'
+
+    # Check if stats are present in DB, if not calculate them
+    not_calculated = force_recalculate
+    stats = db_stats.get_user_artist_map(user['id'], stats_range)
+    if stats is None or getattr(stats, stats_range) is None:
+        not_calculated = True
+
+    # Check if the stats present in DB have been calculated in the past week, if not recalculate them
+    stale = False
+    if not not_calculated:
+        last_updated = getattr(stats, stats_range).last_updated
+        if (datetime.now() - datetime.fromtimestamp(last_updated)).days >= 7:
+            stale = True
+
+    if stale or not_calculated:
+        artist_stats = db_stats.get_user_artists(user['id'], stats_range)
+
+        # If top artists are missing, return the stale stats if present, otherwise return 204
+        if artist_stats is None or getattr(artist_stats, stats_range) is None:
+            if stale:
+                result = stats
+            else:
+                raise APINoContent('')
+        else:
+            # Calculate the data
+            artist_msids = []
+            artist_mbids = []
+            top_artists = getattr(artist_stats, stats_range).artists
+            for artist in top_artists:
+                if artist.artist_msid is not None:
+                    artist_msids.append(artist.artist_msid)
+                else:
+                    artist_mbids += artist.artist_mbids
+
+            country_code_data = _get_country_codes(artist_msids, artist_mbids)
+            result = UserArtistMapStatJson(**{
+                stats_range: {
+                    "artist_map": country_code_data,
+                    "from_ts": int(getattr(artist_stats, stats_range).from_ts),
+                    "to_ts": int(getattr(artist_stats, stats_range).to_ts),
+                    "last_updated": int(datetime.now().timestamp())
+                }
+            })
+
+            # Store in DB for future use
+            db_stats.insert_user_artist_map(user['id'], result)
```

Comment:
What happens if this insert fails? We should catch an exception here and return the data either ways, given that we've calculated it already.

Reply:
Also, let's have a test for this case, where we've calculated the data but the insert fails.